internal/store/generic.go

```diff
@@ -0,0 +1,399 @@
+package store
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/flterrors"
+	"github.com/flightctl/flightctl/internal/store/model"
+	"github.com/flightctl/flightctl/internal/util"
+	"github.com/google/uuid"
+	"github.com/samber/lo"
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+type IntegrationTestCallback func()
+
+// GenericStore provides generic CRUD operations for resources
+// P is a pointer to a model, for example: *model.Device
+// M is the model, for example: model.Device
+// A is the API resource, for example: api.Device
+// ML is the model list, for example: model.DeviceList
+// AL is the API list, for example: api.DeviceList
+type GenericStore[P model.Generic, M any, A any, ML model.GenericList, AL any] struct {
+	db  *gorm.DB
+	log logrus.FieldLogger
+
+	// Conversion functions between API and model types
+	apiToModelPtr   func(*A) (P, error)
+	modelPtrToAPI   func(P, ...model.APIResourceOption) (*A, error)
+	listModelToAPI  func(ML, *string, *int64) (AL, error)
+	modelPtrToModel func(*M) P
+	modelModelToPtr func(P) *M
+
+	// Callback for integration tests to inject logic
+	IntegrationTestCreateOrUpdateCallback IntegrationTestCallback
+}
+
+type Resource struct {
+	Table string
+	OrgID string
+	Name  string
+}
+
+func NewGenericStore[P model.Generic, M any, A any, ML model.GenericList, AL any](
```

Comment:
Isn't P used always as interface? If yes then it is a concrete type and can be removed from the generic parameter list.

Isn't ML just slice of M  ? If yes, can this parameter removed ?